bounswe/bounswe2022group8
# feature/BE-30: Exhibitions

This PR is directly related to #343. Additionally, it involves minor fixes/updates on other parts:
* Date time format is updated. Older version was something like this: `2022-11-18T13:51:56.342042Z`. New version will be something like this: `08-12-2022 00:38:25`.
* Time zone is updated. (It was UTC-0, now it's Istanbul/Europe).
* Minor fixes on the documentation of art item related APIs. 
* Major change: Implementation of exhibition related APIs:
   * Created four new models: `AbstractExhibition`, `VirtualExhibition`, `OfflineExhibition`, `ExhibitionArtItem`. `AbstractExhibition` is an abstract model that doesn't have a correspondence in the database directly. Offline and online exhibitions are derived from it.
   * Implemented all the API endpoints listed in #343. Documented them with Swagger.
   * Implemented unit tests to test helper function I've implemented for the APIs.
* **IMPORTANT**: 
   * Now, we have APIs for both online and offline exhibitions. Offline exhibitions only require a single image (poster) from the user - it also requests some information about the location such as longitude and latitude. Frontend should provide this information using a library or a geolocation API.
   * For online exhibitions, we don't take any location information. During the creation of an online exhibition, user can add images to it by two means: either via uploading from scratch or from his gallery (previously uploaded images). Previously uploaded images should be provided as a list of IDs in the field `artitems_gallery`. New images should be provided as list of base64 encodings in the field `artitems_upload`. During our meetings, we didn't give a final decision at this point, therefore I wanted to provide both of the opportunities to users. (Please keep in mind that uploading more than one image to S3 will take some time - so you may want to add some 'loading' effect or something)
   * Lastly, for now adding collaborators to the exhibition doesn't really mean anything for the online scenario. Because the only time one can upload images to an exhibition is during its creation. I can work on version 2 of exhibition related PRs if we take such requests from frontend and mobile in the future. But, to be realistic, I believe this implementation is more than enough for a majority of the exhibition related requirements we have and it's overkilling to try to implement all those additional features now while we have lots of other works to do in our schedule.)
* How to test:
   * Code may seem a little bit overwhelming - so you don't really need to analyze it in detail. Best you can do is to test the APIs from your Postman. I have a small Postman collection for exhibition related APIs which can be found here:
   [![Run in Postman](https://run.pstmn.io/button.svg)](https://app.getpostman.com/run-collection/37cbcbd48614b64571ae?action=collection%2Fimport)
   *  You can have a very brief look at the Swagger documentation to see If I'm missing something. Sometimes I make simple mistakes while writing those docs.
   * I tried to cover edge cases as much as possible. Try APIs with different inputs. If it's working properly and docs seems fine - then please approve. Thanks!
